assembly/src/assembler/mast_forest_builder.rs

```diff
@@ -347,18 +333,17 @@ impl MastForestBuilder {
 impl MastForestBuilder {
     /// Adds a node to the forest, and returns the [`MastNodeId`] associated with it.
     ///
-    /// If a [`MastNode`] which is equal to the current node was previously added, the previously
-    /// returned [`MastNodeId`] will be returned. This enforces this invariant that equal
-    /// [`MastNode`]s have equal [`MastNodeId`]s.
-    fn ensure_node(&mut self, node: MastNode) -> Result<MastNodeId, AssemblyError> {
-        let node_digest = node.digest();
+    /// Note adding the same [`MastNode`] twice will result in two different [`MastNodeId`]s being
```

Comment:
This isn't true if the MAST nodes are exactly identical right? i.e. same hash, same decorators? I think the only condition where duplicate nodes are required/expected, are when the MAST root is the same, but the decorators differ, otherwise we can collapse duplicates into a single node as before. That way we still largely avoid duplicating a bunch of identical nodes, and only duplicate on an as-needed basis.

Reply:
Oops, this is what the implementation does, but forgot to fix the docstring.

Fixed in #1466